utils/dsp.py

```diff
@@ -47,35 +51,92 @@ def __init__(self,
+            norm="slaney",
             n_mels=self.n_mels,
-            fmin=self.fmin,
-            fmax=self.fmax)
-        if normalize:
-            mel = self.normalize(mel)
-        return mel
+            mel_scale="slaney",
+            f_min=self.fmin,
+            f_max=self.fmax,
+        ).to(self.device)
+
+        return mel_transform
+
+    def load_wav(self, path: Union[str, Path], mono: bool = True) -> torch.Tensor:
+        """Load audio file into a tensor"""
+        effects = []
+        metadata = torchaudio.info(path)
+
+        # merge channels if source is multichannel
+        if mono and metadata.num_channels > 1:
+            effects.extend([
+                ["remix", "-"] # convert to mono
+            ])
+
+        # resample if source sample rate is different from desired sample rate
+        if metadata.sample_rate != self.sample_rate:
+            effects.extend([
+                ["rate", f'{self.sample_rate}'],
+            ])
+
+        waveform, _ = torchaudio.sox_effects.apply_effects_file(path, effects=effects)
+        return waveform
+
+    def save_wav(self, waveform: torch.Tensor, path: Union[str, Path]) -> None:
+        """Save waveform to file"""
+        torchaudio.save(filepath=path, src=waveform, sample_rate=self.sample_rate)
+
+    def adjust_volume(self, waveform: torch.Tensor) -> torch.Tensor:
+        """Adjust volume of the waveform"""
+        return self.volume_transform(waveform)
+
+    def adjust_volume_batched(self, data: List[torch.Tensor]) -> List[torch.Tensor]:
+        """Adjust volume of the waveforms in the batch"""
+        lengths = [tensor.size(1) for tensor in data]
+        padded_batch = [torch.nn.functional.pad(x, (0, max(lengths) - x.size(1))) for x in data]
+        stacked_tensor = torch.stack(padded_batch, dim=0)
+        processed_batch = self.adjust_volume(stacked_tensor)
+        result = [processed_waveform[:, :lengths[index]] for index, processed_waveform in enumerate(processed_batch)]
+        return result
+
+    def waveform_to_mel_batched(self, batch):
+        """Convert waveform to mel spectrogram for the batch of waveforms"""
+        lengths = [tensor.size(1) for tensor in batch]
+        expected_mel_lengths = [x // self.hop_length + 1 for x in lengths]
+        padded_batch = [torch.nn.functional.pad(x, (0, max(lengths) - x.size(1))) for x in batch]
+        batch_tensor = torch.stack(padded_batch, dim=0).to(self.device)
+        mels = self.waveform_to_mel(batch_tensor)
+        list_of_mels = [mel[:, :, :expected_mel_lengths[index]] for index, mel in enumerate(mels)]
+        return list_of_mels
+
+    def waveform_to_mel(self, waveform: torch.Tensor, normalized=True) -> torch.Tensor:
+        """Convert waveform to mel spectrogram"""
+        mel_spec = self.mel_transform(waveform)
+        if normalized:
+            mel_spec = self.normalize(mel_spec)
+        return mel_spec
 
     def griffinlim(self, mel: np.array, n_iter=32) -> np.array:
```

Comment:
I would suggest to use the same type (torch.Tensor) for all functions to be consistent (now some run on np.arrays and some on tensors)